### ISSUE TITLE
FieldHooks for `IWorldElements`

### ISSUE THREAD

**Issue by mpmxyz** [Yellow-Dog-Man/Resonite-Issues#1257]
### Is your feature request related to a problem? Please describe.

We have `ValueFieldHook`s and `ObjectFieldHooks`. But since all `SyncRef<T>` are `IField<RefID>` and not `IField<T>` we can't drive those directly.
`ValueFieldHook<RefID>` works but you don't get OnHook-Events if you write to the reference directly.
resrec:///U-TheAutopilot/R-ed458c46-da26-43a2-9644-d9e509768e4d
![grafik](https://github.com/Yellow-Dog-Man/Resonite-Issues/assets/8538547/841a3ff0-3a9a-4657-971f-99c19a8253ea)
A is `IField<Slot>` which cannot exist.
B is `IFIeld<RefID` and works.
If you enable the drive (C.1) and write to the field (C.2) you will not see any OnHook events. (C.3)

### Describe the solution you'd like

There should be a FieldHook that can drive `SynRef`s directly.

### Describe alternatives you've considered

Using `ValueFieldHook<RefID>` but with a fixed OnHook for object writes.

### Additional Context

This issue has been opened as an enhancement instead of a bug report since I personally consider `SyncRef`s being an `IField<RefID>` an implementation detail. How is the future regarding IWorldElements, RefIDs and fields storing references?
Is this supposed to be a permanent part of Resonite or is #702 potentially changing this?

**Comment by Nytra** [comment]
I guess you are actually looking for a RefHook node, one that would target ISyncRef<T>

Also I don't think it's technically true that all SyncRef<T> are IField<RefID> since you can reference things which are types in any assembly which don't have to be IWorldElement

**Comment by mpmxyz** [comment]
`ISyncRef<T>` is actually restricted to  `T : class, IWorldElement`
But it is true that there is another edge-case: UserRef is an `ISyncRef` but not a `SyncRef` and does not implement `IField` at all.

**Comment by JackTheFoxOtter** [comment]
I ran into the same issue today.

I was trying to create a `FieldHook<User>` node, only to realize it doesn't actually allow a `SyncRef<User>` as an input, which I wasn't  expecting. The only way to achieve this seems to be using `FieldHook<RefID>` as outlined in the alternative (thanks for that!), which does however make the "OnHook" event of the `FieldHook` node rather useless, as I don't think there's a node giving you access to the target reference of a `RefID` input.

**Comment by 3x1t-5tyl3** [comment]
It seems that the entire "ObjectFieldHook<T>" just doesn't work whatsoever as it requires a ValueType and all references are just cast to RefID

**Comment by JackTheFoxOtter** [comment]
No it works, but only for data model value types. Strings for example are value types on the data model, but object types in ProtoFlux. That is what ObjectFieldHook is for.

**Comment by 3x1t-5tyl3** [comment]
> No it works, but only for data model value types. Strings for example are value types on the data model, but object types in ProtoFlux. That is what ObjectFieldHook is for.

Yes but even then IField doesn't "cast" or whatever- idk man I don't do programming- properly to references it casts to RefID's internally; Which *is* a valuetype.

**Comment by JackTheFoxOtter** [comment]
Ah yeah the node doesn't auto-overload to the value variant.
